odev/common/python.py

```diff
@@ -230,12 +228,13 @@ def __pip_install_progress(self, options: str, message: str = "Installing packag
                     buffer.clear()
                     packages = line.split(" ")[2:]
 
-        logger.info(f"Successfully installed {len(packages)} python packages")
-        installed_packages = [
-            f"{string.stylize(name, 'bold color.purple')} == {string.stylize(version, 'bold color.cyan')}"
-            for name, version in (package.rsplit("-", 1) for package in packages)
-        ]
-        logger.debug(f"Installed python packages:\n{string.join_bullet(installed_packages)}")
+        if packages:
```

Comment:
This pr is a good candidate to also change the behavior of the logger call when installing a python package.
Why not print them all instead of using the same log line when the package change ? 

I would like to see : 

```
Package lxml == 4.9.3 Updated
Package MarkupSafe == 2.1.2 Installed
Insalling : pypdf == 4.2 
```

With progress or spinning for the current ongoing line

Reply:
> I would like to see :
> 
> ```
> Package lxml == 4.9.3 Updated
> Package MarkupSafe == 2.1.2 Installed
> Insalling : pypdf == 4.2 
> ```
> 
> With progress or spinning for the current ongoing line

This is not possible to do it dynamically, as the only moment `pip` tells us about what exact package it is working on is during the download phase (example pip output below).

```
Ignoring setuptools: markers 'python_version < "3.8"' don't match your environment
Ignoring setuptools: markers 'python_version >= "3.8" and python_version < "3.12"' don't match your environment
Collecting ipdb (from -r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 1))
  Using cached ipdb-0.13.13-py3-none-any.whl.metadata (14 kB)
Collecting phonenumbers (from -r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 2))
  Using cached phonenumbers-8.13.45-py2.py3-none-any.whl.metadata (10 kB)
Collecting pudb (from -r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 3))
  Using cached pudb-2024.1.2-py3-none-any.whl.metadata (4.4 kB)
Collecting pydevd-odoo (from -r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 4))
  Using cached pydevd_odoo-1.2.2-py3-none-any.whl.metadata (1.5 kB)
Collecting setuptools>59.0.0 (from -r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 7))
  Using cached setuptools-75.1.0-py3-none-any.whl.metadata (6.9 kB)
Collecting websocket-client (from -r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 8))
  Using cached websocket_client-1.8.0-py3-none-any.whl.metadata (8.0 kB)
Collecting ipython>=7.31.1 (from ipdb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 1))
  Using cached ipython-8.27.0-py3-none-any.whl.metadata (5.0 kB)
Collecting decorator (from ipdb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 1))
  Using cached decorator-5.1.1-py3-none-any.whl.metadata (4.0 kB)
Collecting urwid>=2.4 (from pudb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 3))
  Using cached urwid-2.6.15-py3-none-any.whl.metadata (11 kB)
Collecting pygments>=2.7.4 (from pudb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 3))
  Using cached pygments-2.18.0-py3-none-any.whl.metadata (2.5 kB)
Collecting jedi<1,>=0.18 (from pudb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 3))
  Using cached jedi-0.19.1-py2.py3-none-any.whl.metadata (22 kB)
Collecting urwid-readline (from pudb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 3))
  Downloading urwid_readline-0.15.1.tar.gz (9.0 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting packaging>=20.0 (from pudb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 3))
  Using cached packaging-24.1-py3-none-any.whl.metadata (3.2 kB)
Collecting matplotlib-inline (from ipython>=7.31.1->ipdb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 1))
  Using cached matplotlib_inline-0.1.7-py3-none-any.whl.metadata (3.9 kB)
Collecting prompt-toolkit<3.1.0,>=3.0.41 (from ipython>=7.31.1->ipdb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 1))
  Using cached prompt_toolkit-3.0.47-py3-none-any.whl.metadata (6.4 kB)
Collecting stack-data (from ipython>=7.31.1->ipdb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 1))
  Using cached stack_data-0.6.3-py3-none-any.whl.metadata (18 kB)
Collecting traitlets>=5.13.0 (from ipython>=7.31.1->ipdb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 1))
  Using cached traitlets-5.14.3-py3-none-any.whl.metadata (10 kB)
Collecting pexpect>4.3 (from ipython>=7.31.1->ipdb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 1))
  Using cached pexpect-4.9.0-py2.py3-none-any.whl.metadata (2.5 kB)
Collecting parso<0.9.0,>=0.8.3 (from jedi<1,>=0.18->pudb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 3))
  Using cached parso-0.8.4-py2.py3-none-any.whl.metadata (7.7 kB)
Collecting typing-extensions (from urwid>=2.4->pudb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 3))
  Using cached typing_extensions-4.12.2-py3-none-any.whl.metadata (3.0 kB)
Collecting wcwidth (from urwid>=2.4->pudb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 3))
  Using cached wcwidth-0.2.13-py2.py3-none-any.whl.metadata (14 kB)
Collecting ptyprocess>=0.5 (from pexpect>4.3->ipython>=7.31.1->ipdb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 1))
  Using cached ptyprocess-0.7.0-py2.py3-none-any.whl.metadata (1.3 kB)
Collecting executing>=1.2.0 (from stack-data->ipython>=7.31.1->ipdb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 1))
  Using cached executing-2.1.0-py2.py3-none-any.whl.metadata (8.9 kB)
Collecting asttokens>=2.1.0 (from stack-data->ipython>=7.31.1->ipdb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 1))
  Using cached asttokens-2.4.1-py2.py3-none-any.whl.metadata (5.2 kB)
Collecting pure-eval (from stack-data->ipython>=7.31.1->ipdb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 1))
  Using cached pure_eval-0.2.3-py3-none-any.whl.metadata (6.3 kB)
Collecting six>=1.12.0 (from asttokens>=2.1.0->stack-data->ipython>=7.31.1->ipdb->-r /home/avs/repos/odoo-odev/odev/odev/static/requirements.txt (line 1))
  Using cached six-1.16.0-py2.py3-none-any.whl.metadata (1.8 kB)
Using cached ipdb-0.13.13-py3-none-any.whl (12 kB)
Using cached phonenumbers-8.13.45-py2.py3-none-any.whl (2.6 MB)
Using cached pudb-2024.1.2-py3-none-any.whl (87 kB)
Using cached pydevd_odoo-1.2.2-py3-none-any.whl (3.7 kB)
Using cached setuptools-75.1.0-py3-none-any.whl (1.2 MB)
Using cached websocket_client-1.8.0-py3-none-any.whl (58 kB)
Using cached ipython-8.27.0-py3-none-any.whl (818 kB)
Using cached jedi-0.19.1-py2.py3-none-any.whl (1.6 MB)
Using cached packaging-24.1-py3-none-any.whl (53 kB)
Using cached pygments-2.18.0-py3-none-any.whl (1.2 MB)
Using cached urwid-2.6.15-py3-none-any.whl (297 kB)
Using cached decorator-5.1.1-py3-none-any.whl (9.1 kB)
Using cached parso-0.8.4-py2.py3-none-any.whl (103 kB)
Using cached pexpect-4.9.0-py2.py3-none-any.whl (63 kB)
Using cached prompt_toolkit-3.0.47-py3-none-any.whl (386 kB)
Using cached traitlets-5.14.3-py3-none-any.whl (85 kB)
Using cached matplotlib_inline-0.1.7-py3-none-any.whl (9.9 kB)
Using cached stack_data-0.6.3-py3-none-any.whl (24 kB)
Using cached typing_extensions-4.12.2-py3-none-any.whl (37 kB)
Using cached wcwidth-0.2.13-py2.py3-none-any.whl (34 kB)
Using cached asttokens-2.4.1-py2.py3-none-any.whl (27 kB)
Using cached executing-2.1.0-py2.py3-none-any.whl (25 kB)
Using cached ptyprocess-0.7.0-py2.py3-none-any.whl (13 kB)
Using cached pure_eval-0.2.3-py3-none-any.whl (11 kB)
Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Building wheels for collected packages: urwid-readline
  Building wheel for urwid-readline (pyproject.toml) ... done
  Created wheel for urwid-readline: filename=urwid_readline-0.15.1-py3-none-any.whl size=9229 sha256=7a669bb3d9f1af9925669d2a4d75c8f900250743d40775d1af397fa424803d83
  Stored in directory: /home/avs/.cache/pip/wheels/7c/9f/91/6658ca67972b70a9d89fabda634a2941c40025910040d0014b
Successfully built urwid-readline
Installing collected packages: wcwidth, pydevd-odoo, pure-eval, ptyprocess, phonenumbers, websocket-client, typing-extensions, traitlets, six, setuptools, pygments, prompt-toolkit, pexpect, parso, packaging, executing, decorator, urwid, matplotlib-inline, jedi, asttokens, urwid-readline, stack-data, pudb, ipython, ipdb
Successfully installed asttokens-2.4.1 decorator-5.1.1 executing-2.1.0 ipdb-0.13.13 ipython-8.27.0 jedi-0.19.1 matplotlib-inline-0.1.7 packaging-24.1 parso-0.8.4 pexpect-4.9.0 phonenumbers-8.13.45 prompt-toolkit-3.0.47 ptyprocess-0.7.0 pudb-2024.1.2 pure-eval-0.2.3 pydevd-odoo-1.2.2 pygments-2.18.0 setuptools-75.1.0 six-1.16.0 stack-data-0.6.3 traitlets-5.14.3 typing-extensions-4.12.2 urwid-2.6.15 urwid-readline-0.15.1 wcwidth-0.2.13 websocket-client-1.8.0
```

I came up with the following:

- During the download phase, update the spinner with the number of packages downloaded rather than their name and version
- List installed/updated packages and their version at the end of the process
- Show the output logs of `pip` when it failed to installed a package

![image](https://github.com/user-attachments/assets/3dfce609-c220-42a3-913a-a8d4785d51ee)